Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/QrCodeRegistrationStateProcessor.kt

```diff
@@ -0,0 +1,74 @@
+import de.rki.coronawarnapp.coronatest.qrcode.InvalidQRCodeException
+import de.rki.coronawarnapp.coronatest.server.CoronaTestResult
+import de.rki.coronawarnapp.coronatest.type.CoronaTest
+import de.rki.coronawarnapp.exception.ExceptionCategory
+import de.rki.coronawarnapp.exception.TransactionException
+import de.rki.coronawarnapp.exception.http.CwaWebException
+import de.rki.coronawarnapp.exception.reporting.report
+import de.rki.coronawarnapp.submission.SubmissionRepository
+import de.rki.coronawarnapp.ui.submission.ApiRequestState
+import de.rki.coronawarnapp.util.ui.SingleLiveEvent
+import timber.log.Timber
+import javax.inject.Inject
+
+class QrCodeRegistrationStateProcessor @Inject constructor(
+    private val submissionRepository: SubmissionRepository
+) {
+
+    data class RegistrationState(
+        val apiRequestState: ApiRequestState,
+        val test: CoronaTest? = null
+    )
+
+    val showRedeemedTokenWarning = SingleLiveEvent<Unit>()
+    val registrationState = MutableLiveData(RegistrationState(ApiRequestState.IDLE))
+    val registrationError = SingleLiveEvent<CwaWebException>()
+
+    suspend fun startQrCodeRegistration(coronaTestQRCode: CoronaTestQRCode, isConsentGiven: Boolean) =
+        try {
+            registrationState.postValue(RegistrationState(ApiRequestState.STARTED))
+            val coronaTest = submissionRepository.registerTest(coronaTestQRCode)
+            if (isConsentGiven) {
+                submissionRepository.giveConsentToSubmission(type = coronaTestQRCode.type)
+            }
+            checkTestResult(coronaTest.testResult)
+            registrationState.postValue(
+                RegistrationState(
+                    ApiRequestState.SUCCESS,
+                    coronaTest
+                )
+            )
+        } catch (err: CwaWebException) {
+            registrationState.postValue(RegistrationState(ApiRequestState.FAILED))
+            registrationError.postValue(err)
+        } catch (err: TransactionException) {
+            if (err.cause is CwaWebException) {
+                registrationError.postValue(err.cause)
+            } else {
+                err.report(ExceptionCategory.INTERNAL)
+            }
+            registrationState.postValue(RegistrationState(ApiRequestState.FAILED))
+        } catch (err: InvalidQRCodeException) {
+            registrationState.postValue(RegistrationState(ApiRequestState.FAILED))
+            Timber.d("deregisterTestFromDevice()")
+            submissionRepository.removeTestFromDevice(type = coronaTestQRCode.type)
+            showRedeemedTokenWarning.postValue(Unit)
+        } catch (err: Exception) {
+            registrationState.postValue(RegistrationState(ApiRequestState.FAILED))
+            err.report(ExceptionCategory.INTERNAL)
+        }
+
+    private fun checkTestResult(testResult: CoronaTestResult) {
+        if (testResult == CoronaTestResult.PCR_REDEEMED) {
+            throw InvalidQRCodeException()
```

Comment:
Please add some message to the exception? Let us borrow  it from https://github.com/corona-warn-app/cwa-app-android/pull/2894/commits/8083fef55febe73b5fc488c4f114ab65c4137078#diff-58ba20a44ecdb979466f8e5c5038e91330761f46f7834a97e5f8555650975418R118